src/component_selector.jl

```diff
@@ -0,0 +1,371 @@
+`ComponentSelector` contains. Particularly for complex `ComponentSelector`s, this may not
+always be very concise or informative, so in these cases constructing the
+`ComponentSelector` with a custom name is recommended.
+"""
+function default_name end
+
+# Override this if you define a ComponentSelector subtype with no name field
+"""
+Get the name of the `ComponentSelector`. This is either the default name or a custom name
+passed in at creation time.
+"""
+get_name(selector::ComponentSelector) =
+    (selector.name !== nothing) ? selector.name : default_name(selector)
+
+"""
+    get_components(selector, sys; filterby = nothing)
+Get the components of the collection that make up the `ComponentSelector`.
+ - `filterby`: optional filter function to apply after evaluating the `ComponentSelector`
+"""
+function get_components end
+
+"""
+    get_components(filterby, selector, sys)
+Get the components of the collection that make up the `ComponentSelector`.
+ - `filterby`: optional filter function to apply after evaluating the `ComponentSelector`
+"""
+get_components(
+    filterby::Union{Nothing, Function},
+    selector::ComponentSelector,
+    sys::SYSTEM_LIKE,
+) =
+    get_components(selector, sys; filterby = filterby)
+
+"""
+    get_groups(selector, sys; filterby = nothing)
+Get the groups that make up the `ComponentSelector`.
+ - `filterby`: optional filter function to apply after evaluating the `ComponentSelector`
+"""
+function get_groups end
+
+"""
+    get_groups(filterby, selector, sys)
+Get the groups that make up the `ComponentSelector`.
+ - `filterby`: optional filter function to apply after evaluating the `ComponentSelector`
+"""
+get_groups(
+    filterby::Union{Nothing, Function},
+    selector::ComponentSelector,
+    sys::SYSTEM_LIKE,
+) =
+    get_groups(selector, sys; filterby = filterby)
+
+"""
+Use the `groupby` property to get the groups that make up the
+`DynamicallyGroupedComponentSelector`
+"""
+function get_groups(
+    selector::DynamicallyGroupedComponentSelector,
+    sys;
+    filterby = nothing,
+)
+    validate_groupby(selector.groupby)
+    (selector.groupby == :all) && return [selector]
+    (selector.groupby == :each) &&
+        return Iterators.map(make_selector,
+            get_components(selector, sys; filterby = filterby))
+    @assert selector.groupby isa Function
+    components = collect(get_components(selector, sys; filterby = filterby))
+
+    partition_results = (selector.groupby).(components)
+    unique_partitions = unique(partition_results)
+    partition_labels = string.(unique_partitions)
+    # Catch the case where `p1 != p2` but `string(p1) == string(p2)`
+    (length(unique_partitions) == length(unique(partition_labels))) ||
+        throw(ArgumentError("Some partitions have the same name when converted to string"))
+
+    return [
+        make_selector(
+            make_selector.(components[partition_results .== Ref(group_result)])...;
```

Comment:
Consider whether this is too clever for the average user. It will be read many times.

Reply:
Refactored with a helper function and no `Ref`.